source/functional-concepts.md

```diff
@@ -0,0 +1,150 @@
+# Functional Programming in Swift: A Primer
+
+Swift is a hybrid language that allows both objective-oriented and functional code to co-exist.
+
+### First-class functions
+
+First-class functions are *values* that can either take other functions as arguments or return them as results. Also refered to as *higher order functions*.
+
+#### Returning Functions
+
+~~~swift
+typealias Prefixer = String -> String
+
+func prefixerFactory(prefix: String) -> Prefixer {
+// func prefix(prefix: String) -> (String -> String)
+
+  func aPrefixer(word: String) -> String {
+    return prefix + word
+  }
+
+  return aPrefixer
+}
+
+let repeat = prefixerFactory("re-")
+repeat("do")
+~~~
+
+*Note*: You can also return an unnamed function, i.e. a closure.
+
+~~~swift
+func otherPrefixerfactory(prefix: String) -> Prefixer {
+  return { word in
+    return prefix + word
+  }
+}
+~~~
+
+#### Taking Functions as Arguments
+
+~~~swift
+func uppercasePrefixer(word: String, prefixer: Prefixer) -> String {
+  return prefixer(word).uppercaseString
+}
+
+uppercasePrefixer("do", repeat)
+~~~
+
+### Functional Purity
+
+Functions should have no side effects. They do not alter the state around them:
+
+#### Immutability
+
+Functions operate on copies of existing values instead of mutating them.
+
+~~~swift
+struct Employee {
+  let name: String
+  var age: Int
+
+  func incrementAge() -> Employee {
+    return Employee(name: self.name, age: self.age + 1)
+  }
+}
+
+struct OtherEmployee {
+  let name: String
+  var age: Int
+
+  mutating func incrementAge() {
+    age = age + 1
+  }
+}
+~~~
+
+#### Referential transparency
+
+Functions return the same value each time they are invoked with the same arguments.
+
+~~~swift
+import UIKit
+
+func randomMultiplier(count: Int) -> Int {
+  return count * random()
+}
+
+randomMultiplier(1)
+randomMultiplier(1)
+~~~
+
+### Function Composition
+
+Functions can be composed to create more complex functions.
+
+~~~swift
+let someRect = CGRect(origin: CGPointZero, size: CGSize(width: 100, height: 100))
+
+func halfSizeRect(rect: CGRect) -> CGRect {
+  let transform = CGAffineTransformMakeScale(0.5, 0.5)
+  return CGRectApplyAffineTransform(rect, transform)
+}
+
+func rotateRect(rect: CGRect) -> CGRect {
+  let transform = CGAffineTransformMakeRotation(1)
+  return CGRectApplyAffineTransform(rect, transform)
+}
+
+func rotateHalfSizeRect(rect: CGRect) -> CGRect {
+  return rotateRect(scaleRect(rect))
+}
+
+let transformedRect = rotateHalfSizeRect(someRect)
+~~~
```

Comment:
Ah you answer it below. Anyway, the above is not composition, it's just applying functions to values.